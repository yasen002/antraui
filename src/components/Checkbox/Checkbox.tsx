import React from "react";
import { FC, MouseEvent, ReactNode, useEffect, useState } from "react";
import { MdCheckBoxOutlineBlank, MdCheckBox } from "react-icons/md";

type CheckboxColor = "primary" | "secondary" | "default";

type checkboxSize = "small" | "medium" | "large";

type ButtonVariant = "contained" | "outlined" | "text";

type LabelPlacement = "top" | "bottom" | "start" | "end";
interface ICheckboxProps {
  color?: CheckboxColor;
  size?: checkboxSize;
  variant?: ButtonVariant;
  disabled?: boolean;
  children?: ReactNode;
  checked?: boolean;
  icon?: ReactNode;
  checkedIcon?: ReactNode;
  label?: String;
  labelPlacement?: LabelPlacement;
  onClick?: (event: MouseEvent) => void;
}

interface IClickPosition {
  x: number;
  y: number;
}

let counter = 0;

const Checkbox: FC<ICheckboxProps> = ({
  color = "primary",
  size = "medium",
  variant = "contained",
  disabled = false,
  checked = false,
  icon = <MdCheckBoxOutlineBlank />,
  checkedIcon = <MdCheckBox />,
  label = "",
  labelPlacement = "end",
  children,
  onClick,
}) => {
  const [clickPosition, setClickPosition] =
    useState<IClickPosition | null>(null);
  const [rippleArr, setRippleArr] = useState<ReactNode[]>([]);
  const [checkstate, setCheck] = useState(checked);

  const handleClick = (e: MouseEvent) => {
    if (disabled) return;
    const { offsetX, offsetY } = e.nativeEvent;
    setClickPosition({ x: offsetX, y: offsetY });
    onClick?.(e);
  };

  useEffect(() => {
    //add the ripple circle to the rippleArr state
    if (clickPosition !== null) {
      const newRipple = (
        <div
          data-testid="ripple-element"
          key={counter++}
          style={{
            //position
            position: "absolute",
            left: clickPosition.x,
            top: clickPosition.y,
            transform: "translate(-50%,-50%)",
          }}
          className={`checkbox-ripple-${color}-${variant}`}
          onAnimationEnd={() => {
            setRippleArr((prev) => {
              let nextRippleArr = [...prev];
              nextRippleArr.shift();
              return nextRippleArr;
            });
          }}
        ></div>
      );
      setRippleArr((prev) => [...prev, newRipple]);
    }
  }, [clickPosition]);

  const constructLabelClassName: () => string = () => {
    return `checkbox--label-${labelPlacement}`;
  };
  const constructClassName: () => string = () => {
    const colorSizeCls = `checkbox-${color}-${size}`;
    const sizeCls = `checkbox-${size}`;
    return ["checkbox", colorSizeCls, sizeCls].join(" ");
  };
  const constructIconClassName: () => string = () => {
    const colorSizeCls = `checkbox--icon-${color}-${size}`;
    const sizeCls = `checkbox--icon-${size}`;
    const dis = disabled ? `checkbox--icon-${size}-disabled` : "";
    if (disabled) {
      return ["checkbox--icon", sizeCls, dis].join(" ");
    } else {
      return ["checkbox--icon", colorSizeCls, sizeCls, dis].join(" ");
    }
  };

  return (
    <label htmlFor="checkbox" className={`checkbox--container-${labelPlacement}`}>
      <span className="checkbox--content">
        <input
          name="checkbox"
          type="checkbox"
          className={constructClassName()}
          checked={checkstate}
          onChange={() => {
            setCheck(!checkstate);
          }}
          disabled={disabled}
        />
        <span className={constructIconClassName()}>
          {checkstate ? checkedIcon : icon}
        </span>
      </span>
      <span
        className={constructLabelClassName()}
        onClick={() => {
          setCheck(!checkstate);
        }}
      >
        {label}
      </span>
    </label>
  );
};

export default Checkbox;
