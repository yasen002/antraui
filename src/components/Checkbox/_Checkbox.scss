@use "../../scss/variable";
@use "../../scss/mixin";

$color-map: (
  primary: (
    default: variable.$primary-color,
    hover: variable.$primary-color-hover,
    disabled: red,
    textColor: #ffffff,
  ),
  secondary: (
    default: variable.$secondary-color,
    hover: variable.$secondary-color-hover,
    disabled: red,
    textColor: #ffffff,
  ),
  default: (
    default: variable.$default-color,
    hover: variable.$default-color-hover,
    disabled: red,
    textColor: rgba(0, 0, 0, 0.87),
  ),
);

@mixin color-size-class-label($color, $size, $labelPlace) {
  .checkbox-#{$color}-#{$size} {
    cursor: pointer;
    position: absolute;
    opacity: 0;
    margin: 9px;
    padding: 9px;
    z-index: 1;
    @if $size == "small" {
      height: variable.$small-box-size;
      width: variable.$small-box-size;
    } @else if $size == "medium" {
      height: variable.$medium-box-size;
      width: variable.$medium-box-size;
    } @else if $size == "large" {
      height: variable.$large-box-size;
      width: variable.$large-box-size;
    }
  }
  .checkbox-#{$color}-#{$size}:disabled {
    cursor: default;
  }

  .checkbox--icon-#{$color}-#{$size} svg {
    fill: map-get(map-get($color-map, $color), "default");
    padding: 9px;
    @if $size == "small" {
      position: absolute;
      height: variable.$small-box-size;
      width: variable.$small-box-size;
    } @else if $size == "medium" {
      position: absolute;
      height: variable.$medium-box-size;
      width: variable.$medium-box-size;
    } @else if $size == "large" {
      position: absolute;
      height: variable.$large-box-size;
      width: variable.$large-box-size;
    }
  }
  .checkbox--icon-#{$size}-disabled svg {
    fill: variable.$disabled-color;
    padding: 9px;
    @if $size == "small" {
      position: absolute;
      height: 15px;
      width: 15px;
    } @else if $size == "medium" {
      position: absolute;
      height: 18px;
      width: 18px;
    } @else if $size == "large" {
      position: absolute;
      height: 21px;
      width: 21px;
    }
  }
  .checkbox--container-#{$labelPlace} {
    display: flex;
    -webkit-box-align: center;
    align-items: center;
    cursor: pointer;
    vertical-align: middle;
    -webkit-tap-highlight-color: transparent;
    @if $labelPlace == "end" {
      flex-direction: row;
    } @else if $labelPlace == "top" {
      flex-direction: column-reverse;
    } @else if $labelPlace == "start" {
      flex-direction: row-reverse;
    } @else if $labelPlace == "bottom" {
      flex-direction: column;
    }
  }
  .checkbox--content {
    display: inline-flex;
  }
  .checkbox--label-#{$labelPlace} {
    display: inline-flex;
    @if $labelPlace == "end" {

    } @else if $labelPlace == "top" {

    } @else if $labelPlace == "start" {

    } @else if $labelPlace == "bottom" {
      
    }
  }
}

$colors: ("primary", "secondary", "default");
$sizes: ("small", "medium", "large");
$labelPlacement: ("end", "top", "start", "bottom");

@mixin generate-color-size-class-labels {
  @each $size in $sizes {
    @each $color in $colors {
      @each $labelPlace in $labelPlacement {
        @include color-size-class-label($color, $size, $labelPlace);
      }
    }
  }
}

@include generate-color-size-class-labels();

.checkbox span {
  user-select: none;
}

@keyframes rippleEffect {
  0% {
    width: 0px;
    height: 0px;
    opacity: 0.3;
  }
  50% {
    width: 75px;
    height: 75px;
    opacity: 0.1;
  }
  100% {
    width: 150px;
    height: 150px;
    opacity: 0;
  }
}
