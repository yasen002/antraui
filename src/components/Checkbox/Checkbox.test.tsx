import { queryByText, findByText, render, screen, fireEvent, waitFor } from "@testing-library/react"
import React from "react"
import Checkbox from "./Checkbox"


describe("Checkbox test cases", () => {
  test("snapshot test for Checkbox", () => {
    const view = render(<Checkbox>SUBMIT</Checkbox>);
    expect(view.asFragment()).toMatchSnapshot();
  });

  test("Checkbox component should be rendered", () => {
    render(<Checkbox />)
    const checkboxId = screen.getByTestId("large-checkbox")
    expect(checkboxId).toBeInTheDocument();
  });

  test("Checkbox should change its color based on color props", async () => {
    const { rerender } = render(<Checkbox color="primary" />);
    let checkboxEl = await screen.findByRole("checkbox");
    expect(checkboxEl).toHaveClass("checkbox-primary-medium");
    rerender(<Checkbox color="secondary" />);
    checkboxEl = await screen.findByRole("checkbox");
    expect(checkboxEl).toHaveClass("checkbox-secondary-medium");
    rerender(<Checkbox color="default" />);
    checkboxEl = await screen.findByRole("checkbox");
    expect(checkboxEl).toHaveClass("checkbox-default-medium");
  })

    test("Checkbox should change its size based on size props", async () => {
    const { rerender } = render(<Checkbox size="medium" />);
    let checkboxEl = await screen.findByRole("checkbox");
    expect(checkboxEl).toHaveClass("checkbox-medium");
    rerender(<Checkbox size="small" />);
    checkboxEl = await screen.findByRole("checkbox");
    expect(checkboxEl).toHaveClass("checkbox-small");
    rerender(<Checkbox size="large" />);
    checkboxEl = await screen.findByRole("checkbox");
    expect(checkboxEl).toHaveClass("checkbox-large");
  })

    test("Checkbox should be disabled if the disabled props is passed, should be true initially, and false after clicking", async () => {
    const mockClickHandler = jest.fn();
    render(<Checkbox onClick={mockClickHandler} />);
    const checkbox = screen.getByTestId("hidden-checkbox") as HTMLInputElement
    expect(checkbox.checked).toEqual(false)
    fireEvent.click(checkbox);
    expect(checkbox.checked).toEqual(true)
  });

  // test("Checkbox should have rippple animation on click", async () => {
  //   render(<Checkbox>SUBMIT</Checkbox>);
  //   let checkboxEl = await screen.findByRole("checkbox");
  //   let rippleEls = screen.queryAllByTestId("ripple-element");
  //   expect(rippleEls).toHaveLength(0);
  //   fireEvent.click(checkboxEl);
  //   rippleEls = screen.queryAllByTestId("ripple-element");
  //   expect(rippleEls).toHaveLength(1);
  //   fireEvent.click(checkboxEl);
  //   rippleEls = screen.queryAllByTestId("ripple-element");
  //   expect(rippleEls).toHaveLength(2);

  //   rippleEls.forEach(rippleEl => {
  //     fireEvent.animationEnd(rippleEl);
  //   });
  //   rippleEls = screen.queryAllByTestId("ripple-element");
  //   expect(rippleEls).toHaveLength(0);
  // });
})