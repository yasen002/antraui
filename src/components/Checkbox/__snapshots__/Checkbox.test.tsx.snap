// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox test cases snapshot test for Checkbox 1`] = `
<DocumentFragment>
  <label
    class="checkbox--container-end"
    for="checkbox"
  >
    <span
      class="checkbox--content"
    >
      <input
        class="checkbox checkbox-primary-medium checkbox-medium"
        data-testid="hidden-checkbox"
        name="checkbox"
        type="checkbox"
      />
      <span
        class="checkbox--icon checkbox--icon-primary-medium checkbox--icon-medium "
        data-testid="large-checkbox"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
          />
        </svg>
      </span>
    </span>
    <span
      class="checkbox--label-end"
    />
  </label>
</DocumentFragment>
`;
